### ISSUE TITLE
Pre-Training Model and Clarification on QA Dataset

### ISSUE THREAD

**Issue by trisongz** [allenai/longformer#20]
Hi - I'm pretty excited about Longformer and the implications it has for long form NLP!

In the paper, it's outlined that the Pre-Training was conducted in 5 total phases with starting sequence length of 2,048 and ending sequence length of 23,040. For additional LM Pre-Training (English), what would be the best way to continue Pre-Training with additional datasets like C5?

Does the Pre-Training method care about whiteline block text (such as the Shakespeare txt) vs one complete document per line?

In the case of Multi-Lingual and Translation tasks, would it be similar to T5, where you would be able to Translation tasks by fine-tuning, or would it be more effective to have the languages visible during the Pre-Training process for better predictions downstream? (If so, would that essentially require retraining from Phase 1?)

In the cheatsheet for Finetuning QA, there's two additional parameters which are:

```
--wikipedia_dir path/to/evidence/wikipedia/
--web_dir path/to/evidence/web/

```
Would wikipedia_dir be enwiki8 and web_dir be text8?

Last Question - since Longformer uses a custom CUDA implementation during runtime for compiling the functions, would that mean that TPU accelerators would not be able to be used in this implementation?

Thanks!

**Comment by ibeltagy** [comment]
> continue Pre-Training with additional datasets like C5
the 5-phase training was only for the char-lm experiments not the pretrain-finetune experiments

> pretraining text format
you can format your data the way you want depending on your dataset reader

> T5
T5 is an encoder-decoder transformer model (has two selfattention blocks) not an encoder-only model as BERT and RoBERTa. I will be curious to see how it will work if both selattention blocks are replaced with `LongformerSelfAttention`, but this is something we haven't explored

**Comment by safooray** [comment]
@ibeltagy are you planning on releasing your pretraining code including data preprocessing, loading ad copying RoBERTa, and pretraining?

**Comment by ibeltagy** [comment]
@trisongz, now that we have code that doesn't need the custom CUDA kernel, you can try to run it on TPUs.

**Comment by ibeltagy** [comment]
@safooray, our pretraining code uses fairseq with no changes other than the model implementation. We currently don't have plans to release that code because we are assuming it is pretty easy to replicate our setting, especially that Longformer is in the process of being added to [huggingface/transformers](https://huggingface.co/transformers/model_doc/longformer.html) and it will be easy to use their scripts for masked language modeling pretraining. 

That said, adding a script that converts a pretrained model (BERT, RoBERTa, .. others) into their long version is in our [TODO list](https://github.com/huggingface/transformers/pull/4352). The pretraining procedure becomes: 
- run this script to convert RoBERTa into its long version (extends the position embedding with the initialization mentioned in the paper)
- collect some training data on long docs
- use HF MLM training scripts to continue pretraining the model (check hyperparams in the paper)
- make a copy of `query`, `key`, `value` linear projections to create `query_global`, `key_global`, `value_global`

**Comment by ibeltagy** [comment]
@safooray, just added a notebook that replicates our procedure for pretrianing Longformer https://github.com/allenai/longformer/blob/master/scripts/convert_model_to_long.ipynb. It can be applied to other pretrained models to convert them into Long.

**Comment by ibeltagy** [comment]
Looks like all questions here were addressed, so will close this issue. Please feel free to reopen or create a new one if you have more questions.
